File: packages/backend/src/poe2scout/api/economy_cache.py
import asyncio
from collections import defaultdict
from datetime import date, datetime, time, timedelta, timezone
import logging
import random
from typing import Generic, List, TypeVar

from pydantic import BaseModel
from poe2scout.db.repositories import (
    currency_item_repository,
    league_repository,
    price_log_repository,
    unique_item_repository,
)
from poe2scout.db.repositories.models import (
    CurrencyItemExtended,
    PriceLogEntry,
    UniqueItemExtended,
)
from poe2scout.db.repositories.price_log_repository.get_item_daily_stats import (
    GetItemDailyStatsDto,
)
from poe2scout.services.pricing import (
    convert_price_log_matrix_from_base,
    convert_prices_from_base,
)

T = TypeVar("T")

logger = logging.getLogger(__name__)


def get_daily_stat_dates():
    current_date = datetime.now(tz=timezone.utc).date()
    return [current_date - timedelta(days=i) for i in range(7)]


def build_daily_stat_price_logs(
    item_ids: list[int],
    dates: list[date],
    daily_stats: list[GetItemDailyStatsDto],
) -> dict[int, list[PriceLogEntry | None]]:
    results: dict[int, list[PriceLogEntry | None]] = {
        item_id: [None] * len(dates) for item_id in item_ids
    }
    date_indices: dict[date, list[int]] = {}
    for index, stat_date in enumerate(dates):
        date_indices.setdefault(stat_date, []).append(index)

    for stat in daily_stats:
        if stat.item_id not in results:
            continue

        for date_index in date_indices.get(stat.day, []):
            results[stat.item_id][date_index] = PriceLogEntry(
                price=stat.avg_price,
                time=datetime.combine(stat.day, time.min),
                quantity=stat.volume,
            )

    return results


class CacheState(BaseModel, Generic[T]):
    value: List[T]
    expires: datetime


class CacheKey(BaseModel):
    category: str
    league_id: int
    realm_id: int
    game_id: int
    reference_currency: str

    class Config:
        frozen = True


class EconomyCache:
    CurrencyCache: dict[CacheKey, CacheState[CurrencyItemExtended]]
    UniqueCache: dict[CacheKey, CacheState[UniqueItemExtended]]

    CurrencyLocks: dict[CacheKey, asyncio.Lock]

    def __init__(self):
        self.CurrencyCache = {}
        self.UniqueCache = {}
        self.CurrencyLocks = defaultdict(asyncio.Lock)

    async def get_currency_page(
        self,
        league_id: int,
        realm_id: int,
        game_id: int,
        category: str,
        reference_currency: str,
        search: str,
    ) -> List[CurrencyItemExtended]:
        items: List[CurrencyItemExtended]
        cache_key = CacheKey(
            category=category, 
            league_id=league_id, 
            realm_id=realm_id,
            game_id=game_id,
            reference_currency=reference_currency
        )

        cache_entry = self.CurrencyCache.get(cache_key)
        if cache_entry and cache_entry.expires > datetime.now():
            logger.info(f"HitCache for {cache_key}")
            items = self.CurrencyCache[cache_key].value
        else:
            lock = self.CurrencyLocks[cache_key]

            async with lock:
                cache_entry = self.CurrencyCache.get(cache_key)
                if cache_entry and cache_entry.expires > datetime.now():
                    logger.info(f"HitCache (AfterLock) for {cache_key}")
                    items = cache_entry.value
                else:
                    logger.info(f"Cache empty for {cache_key}, fetching from DB.")
                    items = await self.fetch_currency_page(cache_key)

        if search:
            items = [item for item in items if item.text == search]

        return items

    async def get_unique_page(
        self,
        league_id: int,
        realm_id: int,
        game_id: int,
        category: str,
        reference_currency: str,
        search: str,
    ) -> List[UniqueItemExtended]:
        items: List[UniqueItemExtended]

        cache_key = CacheKey(
            category=category, 
            league_id=league_id, 
            realm_id=realm_id,
            game_id=game_id,
            reference_currency=reference_currency
        )
        if (
            self.UniqueCache.get(cache_key) is not None
            and self.UniqueCache[cache_key].expires > datetime.now()
        ):
            logger.info(f"HitCache for {cache_key}")
            items = self.UniqueCache[cache_key].value
        else:
            logger.info(f"Cache empty for {cache_key}")
            items = await self.fetch_unique_page(cache_key)
            self.UniqueCache[cache_key] = CacheState[UniqueItemExtended](
                value=items,
                expires=datetime.now() + timedelta(hours=1, minutes=random.randint(0, 15)),
            )
        if search != "":
            items = [item for item in items if item.name.lower() == search.lower()]

        return items

    async def fetch_unique_page(self, cache_key: CacheKey) -> List[UniqueItemExtended]:
        items_in_current_league, unique_items = await asyncio.gather(
            league_repository.get_items_in_current_league(
                cache_key.league_id,
                cache_key.realm_id,
            ),
            unique_item_repository.get_unique_items_by_category(cache_key.category),
        )
        items_in_current_league_set = set(items_in_current_league)

        unique_items = [
            uniqueItem
            for uniqueItem in unique_items
            if uniqueItem.item_id in items_in_current_league_set
        ]
        item_ids = [item.item_id for item in unique_items]

        if not item_ids:
            return []

        daily_stat_dates = get_daily_stat_dates()
        price_logs = build_daily_stat_price_logs(
            item_ids,
            daily_stat_dates,
            await price_log_repository.get_item_daily_stats(
                item_ids,
                cache_key.league_id,
                cache_key.realm_id,
                daily_stat_dates,
            ),
        )

        reference_currency_price = 1.0
        if cache_key.reference_currency:
            reference_currency_item = await currency_item_repository.get_currency_item(
                cache_key.reference_currency,
                cache_key.game_id,
            )
            if reference_currency_item is None:
                raise Exception()

            reference_currency_price = await price_log_repository.get_item_price(
                reference_currency_item.item_id,
                cache_key.league_id,
                cache_key.realm_id,
                None,
            )
            reference_currency_logs = build_daily_stat_price_logs(
                [reference_currency_item.item_id],
                daily_stat_dates,
                await price_log_repository.get_item_daily_stats(
                    [reference_currency_item.item_id],
                    cache_key.league_id,
                    cache_key.realm_id,
                    daily_stat_dates,
                ),
            )[reference_currency_item.item_id]
            price_logs = convert_price_log_matrix_from_base(
                price_logs,
                reference_currency_logs,
            )

        last_price = dict.fromkeys(item_ids, 0.0)

        prices = await price_log_repository.get_item_prices(
            item_ids, 
            cache_key.league_id,
            cache_key.realm_id
        )

        prices_lookup = {price.item_id: price for price in prices}
        converted_current_prices = convert_prices_from_base(
            {
                price.item_id: float(price.price)
                for price in prices_lookup.values()
            },
            reference_currency_price,
        )

        items = [
            UniqueItemExtended(**item.model_dump(), price_logs=price_logs[item.item_id])
            for item in unique_items
        ]

        for item in items:
            last_price[item.item_id] = converted_current_prices[item.item_id]

        items.sort(
            key=lambda item: (
                last_price[item.item_id] if item.item_id in last_price else 0
            ),
            reverse=True,
        )

        items = [
            UniqueItemExtended(
                **item.model_dump(exclude={"current_price", "current_quantity"}),
                current_price=last_price[item.item_id],
                current_quantity=prices_lookup[item.item_id].quantity,
            )
            for item in items
        ]

        return items

    async def fetch_currency_page(self, cache_key: CacheKey) -> List[CurrencyItemExtended]:
        items_in_current_league, currency_items = await asyncio.gather(
            league_repository.get_items_in_current_league(
                cache_key.league_id,
                cache_key.realm_id,
            ),
            currency_item_repository.get_currency_items_by_category(cache_key.category),
        )
        items_in_current_league_set = set(items_in_current_league)
        currency_items = [
            currencyItem
            for currencyItem in currency_items
            if currencyItem.item_id in items_in_current_league_set
        ]
        item_ids = [item.item_id for item in currency_items]

        if not item_ids:
            next_hour = datetime.now().replace(
                minute=0,
                second=0,
                microsecond=0,
            ) + timedelta(hours=1)

            expiration_time = next_hour + timedelta(minutes=random.randint(5, 10))

            self.CurrencyCache[cache_key] = CacheState[CurrencyItemExtended](
                value=[],
                expires=expiration_time,
            )
            return []

        daily_stat_dates = get_daily_stat_dates()
        price_logs = build_daily_stat_price_logs(
            item_ids,
            daily_stat_dates,
            await price_log_repository.get_item_daily_stats(
                item_ids,
                cache_key.league_id,
                cache_key.realm_id,
                daily_stat_dates,
            ),
        )

        reference_currency_price = 1.0
        if cache_key.reference_currency:
            reference_currency_item = await currency_item_repository.get_currency_item(
                cache_key.reference_currency,
                cache_key.game_id,
            )

            if reference_currency_item is None:
                raise Exception()

            reference_currency_price = await price_log_repository.get_item_price(
                reference_currency_item.item_id,
                cache_key.league_id,
                cache_key.realm_id,
                None,
            )
            reference_currency_logs = build_daily_stat_price_logs(
                [reference_currency_item.item_id],
                daily_stat_dates,
                await price_log_repository.get_item_daily_stats(
                    [reference_currency_item.item_id],
                    cache_key.league_id,
                    cache_key.realm_id,
                    daily_stat_dates,
                ),
            )[reference_currency_item.item_id]
            price_logs = convert_price_log_matrix_from_base(
                price_logs,
                reference_currency_logs,
            )

        last_price = dict.fromkeys(item_ids, 0.0)

        prices = await price_log_repository.get_item_prices(
            item_ids, 
            cache_key.league_id,
            cache_key.realm_id)

        prices_lookup = {price.item_id: price for price in prices}
        converted_current_prices = convert_prices_from_base(
            {
                price.item_id: float(price.price)
                for price in prices_lookup.values()
            },
            reference_currency_price,
        )

        items = [
            CurrencyItemExtended(**item.model_dump(), price_logs=price_logs[item.item_id])
            for item in currency_items
        ]

        for item in items:
            last_price[item.item_id] = converted_current_prices[item.item_id]

        items.sort(
            key=lambda item: (
                last_price[item.item_id] if item.item_id in last_price else 0
            ),
            reverse=True,
        )

        items = [
            CurrencyItemExtended(
                **item.model_dump(exclude={"current_price", "current_quantity"}),
                current_price=last_price[item.item_id],
                current_quantity=prices_lookup[item.item_id].quantity,
            )
            for item in items
        ]

        next_hour = datetime.now().replace(
            minute=0,
            second=0,
            microsecond=0,
        ) + timedelta(hours=1)

        expiration_time = next_hour + timedelta(minutes=random.randint(5, 10))

        self.CurrencyCache[cache_key] = CacheState[CurrencyItemExtended](
            value=items,
            expires=expiration_time,
        )

        return items

File: net/Poe2scout.Api/EconomyCache.cs
using System.Collections.Concurrent;
using Poe2scout.Models;
using Poe2scout.Repositories.CurrencyItem;
using Poe2scout.Repositories.League;
using Poe2scout.Repositories.PriceLog;
using Poe2scout.Repositories.PriceLog.Models;
using Poe2scout.Repositories.UniqueItem;

namespace Poe2scout.Api;

public record CacheKey(
  string Category,
  int LeagueId,
  int RealmId,
  int GameId,
  string ReferenceCurrency,
  int DataPoints,
  int FrequencyHours);

public record CacheEntry<T>(List<T> Value, DateTime ExpiresUtc);

public class EconomyCache(
  ICurrencyItemRepository currencyItemRepository, 
  IUniqueItemRepository uniqueItemRepository,
  ILeagueRepository leagueRepository,
  IPriceLogRepository priceLogRepository)
{
  private readonly Random random = new();
  
  private readonly ConcurrentDictionary<CacheKey, CacheEntry<CurrencyItemExtended>> currencyCache = new();
  private readonly ConcurrentDictionary<CacheKey, CacheEntry<UniqueItemExtended>> uniqueCache = new();
  
  public async Task<List<CurrencyItemExtended>> GetCurrencyPage(
    int leagueId,
    int realmId,
    int gameId,
    string category,
    string referenceCurrency,
    int dataPoints,
    int frequencyHours,
    string? search)
  {
    List<CurrencyItemExtended> result;
    var cacheKey = new CacheKey(category, leagueId, realmId, gameId, referenceCurrency, dataPoints, frequencyHours); 
    var cacheEntry = currencyCache.GetValueOrDefault(cacheKey);

    if (cacheEntry is not null && cacheEntry.ExpiresUtc > DateTime.UtcNow)
    {
      result = cacheEntry.Value;
    }
    else
    {
      result = await FetchCurrencyPage(cacheKey);
    }

    if (!string.IsNullOrEmpty(search))
    {
      result = result.Where(i => i.Text == search).ToList();
    }

    return result;
  }

  public async Task<List<UniqueItemExtended>> GetUniquePage(
    int leagueId,
    int realmId,
    int gameId,
    string category,
    string referenceCurrency,
    int dataPoints,
    int frequencyHours,
    string? search)
  {
    List<UniqueItemExtended> result;
    var cacheKey = new CacheKey(category, leagueId, realmId, gameId, referenceCurrency, dataPoints, frequencyHours); 
    var cacheEntry = uniqueCache.GetValueOrDefault(cacheKey);

    if (cacheEntry is not null && cacheEntry.ExpiresUtc > DateTime.UtcNow)
    {
      result = cacheEntry.Value;
    }
    else
    {
      result = await FetchUniquePage(cacheKey);
      uniqueCache[cacheKey] = new CacheEntry<UniqueItemExtended>(
        result,
        DateTime.UtcNow.AddHours(1).AddMinutes(random.Next(0, 16)));
    }

    if (!string.IsNullOrEmpty(search))
    {
      result = result.Where(i => string.Equals(i.Name, search, StringComparison.OrdinalIgnoreCase)).ToList();
    }

    return result;
  }

  private async Task<List<UniqueItemExtended>> FetchUniquePage(CacheKey cacheKey)
  {
    var leagueItemsTask = leagueRepository.GetItemsInCurrentLeague(cacheKey.LeagueId, cacheKey.RealmId);
    var categoryItemsTask = uniqueItemRepository.GetUniqueItemsByCategory(cacheKey.Category);

    await Task.WhenAll(leagueItemsTask, categoryItemsTask);
    
    var leagueItems = await leagueItemsTask;
    var categoryItems = await categoryItemsTask;
    
    var itemsInCurrentLeague = new HashSet<int>(leagueItems);

    var uniqueItems = categoryItems.Where(c => itemsInCurrentLeague.Contains(c.ItemId)).ToList();
    var itemIds = uniqueItems.Select(i => i.ItemId).ToList();

    if (itemIds.Count == 0)
    {
      return [];
    }
    
    var timeUtc = DateTime.UtcNow;
    var priceLogs = await GetCategoryPriceLogs(
      itemIds,
      cacheKey.LeagueId,
      cacheKey.RealmId,
      cacheKey.DataPoints,
      cacheKey.FrequencyHours,
      timeUtc);

    var referenceCurrencyPrice = 1.0;
    if (!string.IsNullOrEmpty(cacheKey.ReferenceCurrency))
    {
      var referenceCurrencyItem = await currencyItemRepository.GetCurrencyItem(
        cacheKey.ReferenceCurrency,
        cacheKey.GameId);

      if (referenceCurrencyItem is null)
      {
        throw new InvalidOperationException("Reference currency does not exist.");
      }

      referenceCurrencyPrice = await priceLogRepository.GetItemPrice(
        referenceCurrencyItem.ItemId,
        cacheKey.LeagueId,
        cacheKey.RealmId,
        null);

      var referenceCurrencyLogs = (await GetCategoryPriceLogs(
        [referenceCurrencyItem.ItemId],
        cacheKey.LeagueId,
        cacheKey.RealmId,
        cacheKey.DataPoints,
        cacheKey.FrequencyHours,
        timeUtc))[referenceCurrencyItem.ItemId];

      priceLogs = ConvertPriceLogMatrixFromBase(priceLogs, referenceCurrencyLogs);
    }

    var prices = await priceLogRepository.GetItemPrices(itemIds, cacheKey.LeagueId, cacheKey.RealmId);
    var pricesLookup = prices.ToDictionary(price => price.ItemId);
    var convertedCurrentPrices = ConvertPricesFromBase(
      pricesLookup.ToDictionary(kvp => kvp.Key, kvp => kvp.Value.Price),
      referenceCurrencyPrice);

    return uniqueItems
      .Select(item =>
      {
        convertedCurrentPrices.TryGetValue(item.ItemId, out var currentPrice);
        pricesLookup.TryGetValue(item.ItemId, out var itemPrice);

        return new UniqueItemExtended(
          item.UniqueItemId,
          item.ItemId,
          item.IconUrl,
          item.Text,
          item.Name,
          item.CategoryApiId,
          item.ItemMetadata,
          item.Type,
          item.IsChanceable,
          item.IsCurrent,
          priceLogs[item.ItemId],
          currentPrice,
          itemPrice?.Quantity);
      })
      .OrderByDescending(item => item.CurrentPrice ?? 0)
      .ToList();
  }

  private async Task<List<CurrencyItemExtended>> FetchCurrencyPage(CacheKey cacheKey)
  {
    DateTime expirationTimeUtc;
    var leagueItemsTask = leagueRepository.GetItemsInCurrentLeague(cacheKey.LeagueId, cacheKey.RealmId);
    var categoryItemsTask = currencyItemRepository.GetCurrencyItemsByCategory(cacheKey.Category);

    await Task.WhenAll(leagueItemsTask, categoryItemsTask);
    
    var leagueItems = await leagueItemsTask;
    var categoryItems = await categoryItemsTask;
    
    var itemsInCurrentLeague = new HashSet<int>(leagueItems);

    var currencyItems = categoryItems.Where(c => itemsInCurrentLeague.Contains(c.ItemId)).ToList();
    
    var itemIds = currencyItems.Select(i => i.ItemId).ToList();

    if (itemIds.Count == 0)
    {
      var nextHourUtc = DateTime.UtcNow.AddHours(1);
      var nextHourExactUtc = new DateTime(nextHourUtc.Year, nextHourUtc.Month, nextHourUtc.Day, nextHourUtc.Hour, 0, 0, DateTimeKind.Utc);
      expirationTimeUtc = nextHourExactUtc + TimeSpan.FromMinutes(5 + random.NextSingle() * 5);
      currencyCache[cacheKey] = new CacheEntry<CurrencyItemExtended>(Value: [], ExpiresUtc: expirationTimeUtc);

      return [];
    }
    
    var timeUtc = DateTime.UtcNow;
    var priceLogs = await GetCategoryPriceLogs(
      itemIds,
      cacheKey.LeagueId,
      cacheKey.RealmId,
      cacheKey.DataPoints,
      cacheKey.FrequencyHours,
      timeUtc);

    var referenceCurrencyPrice = 1.0;
    if (!string.IsNullOrEmpty(cacheKey.ReferenceCurrency))
    {
      var referenceCurrencyItem = await currencyItemRepository.GetCurrencyItem(
        cacheKey.ReferenceCurrency,
        cacheKey.GameId);

      if (referenceCurrencyItem is null)
      {
        throw new InvalidOperationException("Reference currency does not exist.");
      }

      referenceCurrencyPrice = await priceLogRepository.GetItemPrice(
        referenceCurrencyItem.ItemId,
        cacheKey.LeagueId,
        cacheKey.RealmId,
        null);

      var referenceCurrencyLogs = (await GetCategoryPriceLogs(
        [referenceCurrencyItem.ItemId],
        cacheKey.LeagueId,
        cacheKey.RealmId,
        cacheKey.DataPoints,
        cacheKey.FrequencyHours,
        timeUtc))[referenceCurrencyItem.ItemId];

      priceLogs = ConvertPriceLogMatrixFromBase(priceLogs, referenceCurrencyLogs);
    }

    var prices = await priceLogRepository.GetItemPrices(itemIds, cacheKey.LeagueId, cacheKey.RealmId);
    var pricesLookup = prices.ToDictionary(price => price.ItemId);
    var convertedCurrentPrices = ConvertPricesFromBase(
      pricesLookup.ToDictionary(kvp => kvp.Key, kvp => kvp.Value.Price),
      referenceCurrencyPrice);

    var result = currencyItems
      .Select(item =>
      {
        convertedCurrentPrices.TryGetValue(item.ItemId, out var currentPrice);
        pricesLookup.TryGetValue(item.ItemId, out var itemPrice);

        return new CurrencyItemExtended(
          item.CurrencyItemId,
          item.ItemId,
          item.CurrencyCategoryId,
          item.ApiId,
          item.Text,
          item.CategoryApiId,
          item.IconUrl,
          item.ItemMetadata,
          priceLogs[item.ItemId],
          currentPrice,
          itemPrice?.Quantity);
      })
      .OrderByDescending(item => item.CurrentPrice ?? 0)
      .ToList();

    expirationTimeUtc = GetNextHourExpirationUtc();
    currencyCache[cacheKey] = new CacheEntry<CurrencyItemExtended>(result, expirationTimeUtc);

    return result;
  }

  private async Task<Dictionary<int, IReadOnlyList<PriceLogEntry?>>> GetCategoryPriceLogs(
    List<int> itemIds,
    int leagueId,
    int realmId,
    int dataPoints,
    int frequencyHours,
    DateTime timeUtc)
  {
    if (frequencyHours == 24)
    {
      var dailyStatDates = GetDailyStatDates(dataPoints, timeUtc);
      
      return BuildDailyStatPriceLogs(
        itemIds,
        dailyStatDates,
        await priceLogRepository.GetItemDailyStats(
          itemIds,
          leagueId,
          realmId,
          dailyStatDates));
    }

    var bucketStarts = GetPriceHistoryBucketStarts(dataPoints, frequencyHours, timeUtc);
    return await priceLogRepository.GetItemPriceBucketStats(
      itemIds,
      leagueId,
      realmId,
      bucketStarts,
      frequencyHours);
  }

  private static List<DateOnly> GetDailyStatDates(int dataPoints, DateTime timeUtc)
  {
    var currentDate = DateOnly.FromDateTime(timeUtc);

    var result = new List<DateOnly>();
    for (var i = 0; i < dataPoints; i++)
    {
      result.Add(currentDate.AddDays(-i));
    }

    return result;
  }

  private static List<DateTime> GetPriceHistoryBucketStarts(int dataPoints, int frequencyHours, DateTime timeUtc)
  {
    var currentBlock = new DateTime(
      timeUtc.Year,
      timeUtc.Month,
      timeUtc.Day,
      timeUtc.Hour / frequencyHours * frequencyHours,
      0,
      0,
      DateTimeKind.Utc);

    return Enumerable
      .Range(0, dataPoints)
      .Select(index => currentBlock.AddHours(index * -frequencyHours))
      .ToList();
  }

  private static Dictionary<int, IReadOnlyList<PriceLogEntry?>> BuildDailyStatPriceLogs(List<int> itemIds, List<DateOnly> dates,
    IEnumerable<ItemDailyStats> dailyStats)
  {
    var results = itemIds
      .Select(i => (i, (IReadOnlyList<PriceLogEntry?>)Enumerable.Repeat<PriceLogEntry?>(null, dates.Count).ToList()))
      .ToDictionary(kvp => kvp.i, kvp => kvp.Item2);

    var dateIndices = new Dictionary<DateOnly, List<int>>();
    for (var i = 0; i < dates.Count; i++)
    {
      if (!dateIndices.TryGetValue(dates[i], out var indices))
      {
        indices = [];
        dateIndices[dates[i]] = indices;
      }

      indices.Add(i);
    }

    foreach (var dailyStat in dailyStats)
    {
      if (!results.TryGetValue(dailyStat.ItemId, out var logs) || !dateIndices.TryGetValue(dailyStat.Day, out var indices))
      {
        continue;
      }

      var mutableLogs = (List<PriceLogEntry?>)logs;
      foreach (var index in indices)
      {
        mutableLogs[index] = new PriceLogEntry(
          dailyStat.AvgPrice,
          dailyStat.Day.ToDateTime(TimeOnly.MinValue),
          dailyStat.Volume);
      }
    }

    return results;
  }

  private static Dictionary<int, IReadOnlyList<PriceLogEntry?>> ConvertPriceLogMatrixFromBase(
    Dictionary<int, IReadOnlyList<PriceLogEntry?>> priceLogsByItemId,
    IReadOnlyList<PriceLogEntry?> referenceCurrencyLogs)
  {
    var convertedLogs = new Dictionary<int, IReadOnlyList<PriceLogEntry?>>();

    foreach (var (itemId, priceLogs) in priceLogsByItemId)
    {
      var convertedItemLogs = new List<PriceLogEntry?>();
      for (var i = 0; i < priceLogs.Count; i++)
      {
        var priceLog = priceLogs[i];
        var referenceLog = i < referenceCurrencyLogs.Count ? referenceCurrencyLogs[i] : null;

        if (priceLog is null || referenceLog is null || referenceLog.Price == 0)
        {
          convertedItemLogs.Add(null);
          continue;
        }

        convertedItemLogs.Add(new PriceLogEntry(
          priceLog.Price / referenceLog.Price,
          priceLog.Time,
          priceLog.Quantity));
      }

      convertedLogs[itemId] = convertedItemLogs;
    }

    return convertedLogs;
  }

  private static Dictionary<int, double> ConvertPricesFromBase(
    Dictionary<int, double> pricesByItemId,
    double referenceCurrencyPrice)
  {
    if (referenceCurrencyPrice == 0)
    {
      return pricesByItemId.ToDictionary(kvp => kvp.Key, _ => 0.0);
    }

    return pricesByItemId.ToDictionary(kvp => kvp.Key, kvp => kvp.Value / referenceCurrencyPrice);
  }

  private DateTime GetNextHourExpirationUtc()
  {
    var nextHourUtc = DateTime.UtcNow.AddHours(1);
    var nextHourExactUtc = new DateTime(
      nextHourUtc.Year,
      nextHourUtc.Month,
      nextHourUtc.Day,
      nextHourUtc.Hour,
      0,
      0,
      DateTimeKind.Utc);
    
    return nextHourExactUtc + TimeSpan.FromMinutes(5 + random.NextSingle() * 5);
  }
}
